### PR TITLE
fix: suback Error Codes Handling

### PR DESCRIPTION
**Pull request by tsteinholz**
- Add error handling for failed suback requests
  - specified constant the suback return code value
  - copied error object used in other packet handlers, adjusted for this packet
  - added error callback method used in other packet handlers
  - passed error object to the callback method

Now the error creates the expected Exception with detail and is available from all error callbacks, consistent with the rest of the library.

### Subscribing to Unauthorized topic

```
error: Error: Subscribe error: Not authorized
    at handleAck (/Users/tsteinholz/Development/MQTT.js/build/lib/handlers/ack.js:102:27)
    at handle (/Users/tsteinholz/Development/MQTT.js/build/lib/handlers/index.js:36:31)
    at work (/Users/tsteinholz/Development/MQTT.js/build/lib/client.js:221:40)
    at writable._write (/Users/tsteinholz/Development/MQTT.js/build/lib/client.js:246:13)
    at writeOrBuffer (/Users/tsteinholz/Development/MQTT.js/node_modules/readable-stream/lib/internal/streams/writable.js:334:12)
    at _write (/Users/tsteinholz/Development/MQTT.js/node_modules/readable-stream/lib/internal/streams/writable.js:283:10)
    at Writable.write (/Users/tsteinholz/Development/MQTT.js/node_modules/readable-stream/lib/internal/streams/writable.js:286:10)
    at TLSSocket.ondata (node:internal/streams/readable:1007:22)
    at TLSSocket.emit (node:events:519:28)
    at addChunk (node:internal/streams/readable:559:12) {
  code: 135
}
granted: [
  {
    topic: 'topic/user/does/NOT/have/access/to,
    qos: 1,
    nl: false,
    rap: false,
    rh: 0,
    properties: undefined
  }
]
```

### Subscribing to authorized topic

```
error: undefined
granted: [
  {
    topic: 'topic/user/has/access/to',
    qos: 1,
    nl: false,
    rap: false,
    rh: 0,
    properties: undefined
  }
]
```

### CODE REVIEW

File: src/lib/handlers/ack.ts [hunk 1]
```diff
@@ -117,19 +117,26 @@ const handleAck: PacketHandler = (client, packet) => {
...
 					// suback with Failure status
 					const topics = client.messageIdToTopic[messageId]
 					if (topics) {
 						topics.forEach((topic) => {
 							delete client['_resubscribeTopics'][topic]
 						})
 					}
+					client['_removeOutgoingAndStoreMessage'](messageId, () => {
+						cb(err, packet)
+					})
```
Comment (robertsLando) [inline src/lib/handlers/ack.ts:134]: In this way cb is called twice once here and once later

**Review by robertsLando — CHANGES_REQUESTED** [review]
Also please add a unit test for this

**Comment by tsteinholz** [comment]
The linter seems to have a problem with passing the error object to the callback

> Error:   132:58  error  Function declared in a loop contains unsafe references to variable(s) 'err'  @typescript-eslint/no-loop-func
https://github.com/mqttjs/MQTT.js/actions/runs/9501359758/job/26187059098?pr=1887

Which might be why this error was not passed in the first place.

I am not sure why this is an error, given error is defined [at the beginning of the method](https://github.com/mqttjs/MQTT.js/blob/main/src/lib/handlers/ack.ts#L57). Do we need to explicitly define this as `null` instead of leaving it implicitly `undefined`?

File: src/lib/handlers/ack.ts (hunk 1 above)
Comment (tsteinholz) [reply src/lib/handlers/ack.ts]: removed

**Review by robertsLando — CHANGES_REQUESTED** [review]
Looks good now, please add a test for this and then I will merge

**Comment by tsteinholz** [comment]
@robertsLando It seems like there is test coverage for the immediate subscribe method and the validation of the topic strings. This may be redundant as there is [a return code for the MQTT Broker to do this validation and reject the subscription request as well.
](https://github.com/mqttjs/MQTT.js/blob/main/src/lib/handlers/ack.ts#L29).

I think I found similar tests in `describe('topic validations when subscribing'` to base the suback return code validations on, however, it does not appear there is any ACL enforced on this demo broker.

```js
		it('should return an error (via callbacks) for topic it does not have access to', function _test(t, done) {
			const client = connect()
			client.subscribe('$SYS/#', (subErr) => {
				client.end(true, (endErr) => {
					if (subErr) {
						return done(endErr)
					}
					done(new Error('Suback errors do NOT work'))
				})
			})
		})
```

Since the topic filters are validated on the client side and don't get sent to the broker, I cannot set an invalid topic string to initate an error subnack from the broker. Other possible suback errors are listed here: https://github.com/mqttjs/MQTT.js/issues/1886

What is the recommended way to emulate the suback packet or set an Access Control List (ACL) for the test broker?

```
Packet {
  cmd: 'suback',
  retain: false,
  qos: 0,
  dup: false,
  length: 3,
  topic: null,
  payload: null,
  granted: [ 0 ],
  messageId: 29845
}
2024-06-13T16:06:38.820Z mqttjs connecting to an MQTT broker...
2024-06-13T16:06:38.820Z mqttjs calling streambuilder for mqtt
    ✖ should return an error (via callbacks) for topic it does not have access to (1.584458ms)
      Error: Suback errors do NOT work
          at <anonymous> (/Users/tsteinholz/Development/MQTT.js/test/abstract_client.ts:549:11)
          at <anonymous> (/Users/tsteinholz/Development/MQTT.js/src/lib/client.ts:1454:7)
          at Store.close (/Users/tsteinholz/Development/MQTT.js/src/lib/store.ts:163:4)
          at <anonymous> (/Users/tsteinholz/Development/MQTT.js/src/lib/client.ts:1446:24)
          at Store.close (/Users/tsteinholz/Development/MQTT.js/src/lib/store.ts:163:4)
          at closeStores (/Users/tsteinholz/Development/MQTT.js/src/lib/client.ts:1445:23)
          at process.processTicksAndRejections (node:internal/process/task_queues:77:11)
```

**Comment by tsteinholz** [comment]
I was able to find what I was looking for from the `describe('with alternate server client'` test cases

**Review by robertsLando — CHANGES_REQUESTED** [review]
I think you should also check that the `error` event to be emitted as it was not before

**Review by robertsLando — CHANGES_REQUESTED** [review]
@tsteinholz It's good now just remember to fix lint issues `npm run lint-fix`

File: test/abstract_client.ts [hunk 2]
```diff
@@ -3975,6 +3975,37 @@ export default function abstractTest(server, config, ports) {
...
+						})
+					})
+				})
+
+				client.subscribe('$SYS/#', (subErr) => {
+					client.end(true, (endErr) => {
+						if (subErr) {
+							assert.strictEqual(
+								subErr.message,
+								'Error: Subscribe error: Not authorized',
```
Comment (robertsLando) [inline test/abstract_client.ts:4000]:
```suggestion
								'Subscribe error: Not authorized',
```

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: test/abstract_client.ts (hunk 2 above)
Comment (robertsLando) [reply test/abstract_client.ts:4000]: The test fails otherwise

Comment (tsteinholz) [reply test/abstract_client.ts]: 0kay, I was having trouble running this test locally. Thanks

**Comment by tsteinholz** [comment]
Seems like the new unit test is passing:
* https://github.com/mqttjs/MQTT.js/actions/runs/9548768742/job/26317563967?pr=1887#step:6:390
* https://github.com/mqttjs/MQTT.js/actions/runs/9548768742/job/26317564394?pr=1887#step:6:366

Not exactly sure why the unit test suite are timing out
* https://github.com/mqttjs/MQTT.js/actions/runs/9548768742/job/26317563967?pr=1887#step:6:877
* https://github.com/mqttjs/MQTT.js/actions/runs/9548768742/job/26317564394?pr=1887#step:6:798

Seems like there is some side-effects added in the test suite. Is there a timeout configured in the test and adding the new test is making it take too long to execute?

**Comment by tsteinholz** [comment]
Increased timeout from 60 seconds / 1 minute to 90 seconds / 1 minute and 30 seconds.

**Comment by robertsLando** [comment]
@tsteinholz it's not a timeout problem it's tests that are flacky, could you remove the change to timeout please? When I notice them are failing because of their "flackyness" I sumply re-run them

**Comment by robertsLando** [comment]
@tsteinholz test was wrong, because it was running on both v5 and v3.1 tests and that reason codes are only supported in v5

**Comment by tsteinholz** [comment]
@robertsLando While reason codes are only supported in the v5 specification. The v3 specification still returns an error, it's just always a "Unspecified Error".

The v3 test would just need to be validating that 0x80 returns an "Unspecified Error".

http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718068

**Comment by robertsLando** [comment]
@tsteinholz Yeah I know that but if I return 135 in granted the mqtt parser doens't like the packet and rejects it with protocol 3/4 so I simply prefer to run this in mqtt v5 tests

**Pull request merged**